### DISCUSSION TITLE
AOC 2022 Day Six 📻

### DISCUSSION THREAD

**Discussion by doleraj** [Virtual-Coffee/virtualcoffee.io#761] [Advent of Code 2022]
**Discussion of solutions for [Advent of Code Day 6](https://adventofcode.com/2022/day/6) here.**

---
Don't forget about the VC Advent of Code private leaderboard:

Go to [leaderboard/private](https://adventofcode.com/2022/leaderboard/private)
Use join code 363496-e6f42283

**Comment by doleraj** [comment]
https://github.com/doleraj/adventofcode/blob/main/aoc2022/src/day06/index.ts

I left my Dumb As Bricks™ part 1 unrefactored to prove I am a Real Developer: I write the dumb thing first and refactor later.

**Reply by turingtestable** [reply]
I did the same thing for part 1 since it seemed straightforward and I didn't want to look up set or uniqueness for an array, but fixed it with a commit. I still have a record of my Dumb as Bricks solution in commit history though. 

I sort of like refactoring rather than trying to anticipate the specific way they are going to change things, whenever I try to be smart they tweak it a different way than i expect for part 2.

**Reply by clandau** [reply]
yea whenever I try to anticipate I'm wrong. today I lucked out.

**Reply by doleraj** [reply]
Yeah, sometimes I'll go back and change part 1 but this year is the year of No Effort.

**Comment by clandau** [comment]
https://github.com/clandau/advent-of-code/blob/main/2022/day-6/index.js

**Reply by tkshill** [reply]
Your code is always so accessible and easy to read. You must have the best PRs!

**Reply by bogdaaamn** [reply]
Very nice, I love it, I never remember how to use substrings (or work on string usually), I'd rather go for arrays, but also today I used slice

**Comment by danieltott** [comment]
Day 6! This was a fun little one 😁 

https://github.com/danieltott/aoc2022/blob/main/src/day06/index.ts

**Reply by danieltott** [reply]
Still trying to figure out if I need to even do a `split` on the input. I _think_ it's faster - otherwise I'd have to run `split` on 4 (or 14) letters every time...

**Reply by clandau** [reply]
🤔 I did it the other way but now I'm curious

**Reply by danieltott** [reply]
@aishwarya-mali [figured it out](https://github.com/aishwarya-mali/advent-of-code/blob/main/2022/Day-6/script.js#L5-L6)! You can feed a string to `new Set()` and it'll split it up on it's own: 
```js
const chars = input.slice(i, i + 4)
const uniqueSet = [...new Set(chars)]
```

**Reply by doleraj** [reply]
wait REALLY?!

**Reply by aishwarya-mali** [reply]
I think it is because of spread operator.

**Reply by danieltott** [reply]
No you don't even need the spread operator!
![image](https://user-images.githubusercontent.com/360261/205958563-eb45053c-0dd2-42c7-88f7-753adc865287.png)

**Reply by danieltott** [reply]
Oh this is wild - using a string and doing `substr` is actually noticeably _slower_ than splitting the original string and using `slice`. 

(new times in green)
![image](https://user-images.githubusercontent.com/360261/205960274-646c4362-d848-49bd-a675-a07efe1fedc3.png)



```diff
-const parseInput = (rawInput: string) => rawInput.split("");
+const parseInput = (rawInput: string) => rawInput;
 
-function findCode(input: string[], codeLength: number) {
+function findCode(input: string, codeLength: number) {
   let i = 0;
   let found;
   while (!found) {
-    if (new Set(input.slice(i, i + codeLength)).size === codeLength) {
+    if (new Set(input.substring(i, i + codeLength)).size === codeLength) {
       found = i;
     }
     i++;
   }
 
   return found + codeLength;
 }

```

**Reply by aishwarya-mali** [reply]
> No you don't even need the spread operator! ![image](https://user-images.githubusercontent.com/360261/205958563-eb45053c-0dd2-42c7-88f7-753adc865287.png)

Ohh yes I thought it in context with split('') to get array.

**Reply by doleraj** [reply]
That makes sense to me. I always assume string stuff is going to be slower than raw arrays.

**Reply by bogdaaamn** [reply]
Nice as ush! Almost the same, but I went for for-loops

**Comment by turingtestable** [comment]
Here's my: [Ruby Solution](https://github.com/turingtestable/AdventOfCode2022/blob/master/challenge6/solution.rb). I did the naive solution first, and then fixed it, feel free to check out the previous commit. This seems like a way easier challenge than the past few days.

**Reply by tkshill** [reply]
Excellent

**Comment by davidalpert** [comment]
here's my F# solution
- [Tuning Trouble](https://github.com/davidalpert/advent-of-code-da/blob/2022/2022/day06_tuning_trouble.fs)

I made deliberate choices about when I used `Seq` (which is somewhat lazy and often iterates on demand) and `Array` to avoid dealing with the latter parts of the string and bail out of the loop as soon as we found a match.

**Reply by tkshill** [reply]
I am so jealous of this solution. Lovely! I totally forgot about `windowed`

**Reply by tkshill** [reply]
I'm thinking you might have been able to use `Seq.indexed` here (if you preferred) and maybe streamline some of the index stuff

Something like?
```
buffer.Trim().ToCharArray()
            |> Seq.windowed n
            |> Seq.indexed
            |> Seq.find (fun value -> snd value |> window |> Array.distinct |> Array.length = n)
            |> fst
            |> ((+) 1)
```

**Comment by aishwarya-mali** [comment]
JS: https://github.com/aishwarya-mali/advent-of-code/blob/main/2022/Day-6/script.js

**Reply by danieltott** [reply]
OMG this is awesome! I had no idea you could feed a string to `Set` and it would use the individual parts!

**Reply by tkshill** [reply]
So elegant!

**Reply by bogdaaamn** [reply]
Amazing! Clean and short, I love it. I also went for a Set and it didn't disappoint

**Comment by tkshill** [comment]
I'm genuinely pissed at how straightforward this was compared to 5

[Solution](https://github.com/tkshill/AdventOfCode2022/blob/main/AOC2022/Day06.fs) using **RECURSION**

**Reply by clandau** [reply]
I'm happy. I can focus on work today 😆

**Reply by bogdaaamn** [reply]
NICE, I _started_ with recursion as well. But I dropped it for my own sanity and went classic

**Comment by danimeyer** [comment]
Ruby: https://github.com/danimeyer/advent-of-code/blob/main/2022/6.rb

I also refactored, but you can find my original solution [here](https://github.com/danimeyer/advent-of-code/commit/783ab704444ced5ceaa66331b239ac3f1e80e136). It's ugly 😂

**Comment by Cerchie** [comment]
finished part 1, will pick up part 2 tomorrow! https://github.com/Cerchie/adventofcode2022/tree/main/day-06

**Reply by tkshill** [reply]
`letterBeforeBefore` and `letterAfterAfter` are my new favourite variable names

**Reply by bogdaaamn** [reply]
ahahaha amazing

**Comment by bogdaaamn** [comment]
Late to the game but here we gooo https://github.com/BogDAAAMN/advent-of-code-2022/blob/main/day-06.js

I don't like too much the for i++ loop solution, but I think I dislike more while(found), so here we go 😂 

But I am proud of myself for reminding Sets exist 😂
